### PR TITLE
Various testutil improvements

### PR DESCRIPTION
**Pull request by thaJeztah**
See individual commits for a description of the changes

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @tiborvass @SamWhited @kolyshkin ptal

File: hack/make/.integration-daemon-stop [hunk 1]
```diff
@@ -11,7 +11,9 @@ if [ ! "$(go env GOOS)" = 'windows' ]; then
 			echo >&2 "warning: PID $pid from $pidFile had a nonzero exit code"
 		fi
 		root=$(dirname "$pidFile")/root
-		umount "$root" || true
+		if [ -d "$root" ]; then
+			umount "$root" || true
```
Comment (thaJeztah) [inline hack/make/.integration-daemon-stop]: in practice, this only finds the `docker.pid` of the daemon that was started by `.integration-daemon-start`, but that daemon uses the default (`/var/lib/docker`) as path, so nothing gets unmounted. I think this will only do something if a daemon was not killed as part of a test

**Comment by thaJeztah** [comment]
Whoops; made a typo; fixed (also found some more small changes)

```
testutil/daemon/daemon.go:95: printf: Wrapf format % is missing verb at end of string (govet)
  		return nil, errors.Wrapf(err, "failed to create daemon socket root %", SockRoot)
```

File: testutil/daemon/daemon.go [hunk 2]
```diff
@@ -382,6 +389,9 @@ func (d *Daemon) StartWithBusybox(t testing.TB, arg ...string) {
 // Kill will send a SIGKILL to the daemon
 func (d *Daemon) Kill() error {
 	if d.cmd == nil || d.Wait == nil {
+		if err := mount.Unmount(d.Root); err != nil {
+			d.log.Logf("[%s] failed to unmount daemon root (%s): %v", d.id, d.Root, err)
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply testutil/daemon/daemon.go]: Dropped this commit, because this would also unmount the daemon root if live-restore is enabled, probably causing this failure:

```
 === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.90s)
     --- FAIL: TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore (2.90s)
         daemon.go:26: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestDockerDaemonSuite/TestExecWithUserAfterLiveRestore"
         docker_cli_daemon_test.go:2723: assertion failed: expression is false: err == nil: Output: unable to find user test: no matching entries in passwd file
```

File: hack/make/.integration-daemon-stop (hunk 1 above)
Comment (kolyshkin) [reply hack/make/.integration-daemon-stop]: I'd rephrase as
```shell
test -d "$root" && umount "$root" || true
```

Comment (kolyshkin) [reply hack/make/.integration-daemon-stop]: and even then you can still get "not mounted" error (afaik there's no way to work around it)

File: testutil/daemon/daemon_unix.go [hunk 3]
```diff
@@ -13,17 +13,17 @@ import (
...
 	t.Helper()
 	// Cleanup network namespaces in the exec root of this
 	// daemon because this exec root is specific to this
 	// daemon instance and has no chance of getting
 	// cleaned up when a new daemon is instantiated with a
 	// new exec root.
-	netnsPath := filepath.Join(execRoot, "netns")
+	netnsPath := filepath.Join(d.execRoot, "netns")
 	filepath.Walk(netnsPath, func(path string, info os.FileInfo, err error) error {
 		if err := unix.Unmount(path, unix.MNT_DETACH); err != nil && err != unix.EINVAL && err != unix.ENOENT {
```
Comment (kolyshkin) [inline testutil/daemon/daemon_unix.go]: While at it, we can switch to using `mount.Unmount()` which does all the same things.

Yet better, use `RecursiveUnmount()` to further simplify things here.

File: testutil/daemon/daemon.go [hunk 4]
```diff
@@ -246,11 +247,15 @@ func (d *Daemon) StartWithLogFile(out *os.File, providedArgs ...string) error {
 		return errors.Wrapf(err, "[%s] could not find docker binary in $PATH", d.id)
 	}
 
+	if d.pidFile == "" {
+		d.pidFile = fmt.Sprintf("%s/docker.pid", d.Folder)
```
Comment (kolyshkin) [inline testutil/daemon/daemon.go]: To me, `path.Join(d.Folder, "docker.pid")` looks more elegant than this `Sprintf`.

File: hack/make/.integration-daemon-stop (hunk 1 above)
Comment (thaJeztah) [reply hack/make/.integration-daemon-stop]: Ah, yes, that'd work

I realize it won't fix all situations (always gonna be a race), but currently, it _always_ prints this message

File: testutil/daemon/daemon.go (hunk 4 above)
Comment (thaJeztah) [reply testutil/daemon/daemon.go]: 🤦‍♂ yes, should've seen that 👍

File: testutil/daemon/daemon.go [hunk 5]
```diff
@@ -395,7 +400,10 @@ func (d *Daemon) Kill() error {
 		return err
 	}
 
-	return os.Remove(fmt.Sprintf("%s/docker.pid", d.Folder))
+	if d.pidFile != "" {
```
Comment (kolyshkin) [inline testutil/daemon/daemon.go]: Would be good to note in the commit message that this fixes bogus error from removing `docker.pid` that was never created.

File: testutil/daemon/daemon_unix.go (hunk 3 above)
Comment (thaJeztah) [reply testutil/daemon/daemon_unix.go]: I now recall I had a PR that made changes here; https://github.com/moby/moby/pull/36511, but for some reason it kept failing.

Let me try using the RecursiveUnmount() that sounds like a clean approach (if it works)

File: testutil/daemon/daemon.go [hunk 6]
```diff
@@ -519,12 +524,7 @@ out2:
 		return err
 	}
 
-	d.cmd.Wait()
-
-	if d.pidFile != "" {
-		return os.Remove(d.pidFile)
-	}
-	return nil
+	return d.cmd.Wait()
```
Comment (kolyshkin) [inline testutil/daemon/daemon.go]: Not quite related to this patch (except see below), but I see that `StartWithLogFile()` already creates a goroutine that calls `d.cmd.Waits()` and sends its return to the channel, `d.Wait`.

Now, this code calls `d.cmd.Wait()` one more time, I wonder why, one wait should be just enough. Looking into git logs, it comes all the way from ancient history. Maybe it's time to fix it. I would remove this second wait.

Now that you're returning the second `d.cmd.Wait()` exit code, it might actually return an error (I haven't checked that, just a hunch) as it was already been waited for.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
left a few nitpicks and one important comment about double `d.cmd.Wait()`

File: testutil/daemon/daemon_unix.go (hunk 3 above)
Comment (kolyshkin) [reply testutil/daemon/daemon_unix.go]: I take that back. Neither `filepath.Walk()` nor `RecursiveUnmount()` should be used here, as per https://github.com/moby/moby/pull/36511#discussion_r172975216

Comment (kolyshkin) [reply testutil/daemon/daemon_unix.go]: I.e. the approach in #36511 is the correct one; I would take that one here )

File: hack/make/.integration-daemon-stop (hunk 1 above)
Comment (thaJeztah) [reply hack/make/.integration-daemon-stop]: hm, so shellcheck doesn't like your approach, and would require a `//nolint` for that; https://github.com/koalaman/shellcheck/wiki/SC2015

I'll keep the current approach (although it's slightly more verbose)

File: testutil/daemon/daemon_unix.go (hunk 3 above)
Comment (thaJeztah) [reply testutil/daemon/daemon_unix.go]: Let me rebase the other one, to keep this one simpler (and in case it's still problematic). I can rebase it on top of this one

File: testutil/daemon/daemon.go [hunk 7]
```diff
@@ -451,12 +461,17 @@ func (d *Daemon) StopWithError() (err error) {
...
 	defer func() {
-		if err == nil {
-			d.log.Logf("[%s] Daemon stopped", d.id)
+		if err != nil {
+			d.log.Logf("[%s] error while stopping daemon: %v", d.id, err)
 		} else {
-			d.log.Logf("[%s] Error when stopping daemon: %v", d.id, err)
+			d.log.Logf("[%s] daemon stopped", d.id)
+			if d.pidFile != "" {
+				_ = os.Remove(d.pidFile)
```
[the remark this answers is not in the document]
Comment (SamWhited) [reply testutil/daemon/daemon.go]: It could potentially let us know that something is wrong with our tests. I'd say that it's certainly not worth failing tests over, but is worth logging personally, but I defer to your judgement.

File: hack/make/.integration-daemon-stop (hunk 1 above)
Comment (kolyshkin) [reply hack/make/.integration-daemon-stop]: Ahh, this is the case when shellcheck is making things worse (and yes, I would not want to add `// nolint`).

Let it stay as it is.

File: testutil/daemon/daemon.go [hunk 8]
```diff
@@ -231,7 +234,7 @@ func (d *Daemon) Start(t testing.TB, args ...string) {
 func (d *Daemon) StartWithError(args ...string) error {
 	logFile, err := os.OpenFile(filepath.Join(d.Folder, "docker.log"), os.O_RDWR|os.O_CREATE|os.O_APPEND, 0600)
 	if err != nil {
-		return errors.Wrapf(err, "[%s] Could not create %s/docker.log", d.id, d.Folder)
+		return errors.Wrapf(err, "[%s] failed to create %s/docker.log", d.id, d.Folder)
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply testutil/daemon/daemon.go]: Actually, `os.OpenFile()` should have the file name in its error message, so printing it explictly is redundant. Just
```go
errors.Wrap(err, "[%s] failed to create logfile")
```
should be enough here.

[the remark this answers is not in the document]
Comment (thaJeztah) [reply testutil/daemon/daemon.go]: Good point, yes I can remove it

File: testutil/daemon/daemon.go (hunk 5 above)
Comment (kolyshkin) [reply testutil/daemon/daemon.go]: @SamWhited I would 
- not care about error returned from `os.Remove()` at all
- not return any error here as this will fail the test

I.e. current approach looks fine for me.

**Review by kolyshkin — APPROVED** [review]
LGTM

File: testutil/daemon/daemon.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply testutil/daemon/daemon.go]: done

File: testutil/daemon/daemon.go (hunk 6 above)
Comment (thaJeztah) [reply testutil/daemon/daemon.go]: done

**Review by SamWhited — APPROVED** [review]

**Comment by thaJeztah** [comment]
Argh; Windows failed, but for some reason, Windows RS5 never prints the actual failure

```

[2019-10-10T18:20:14.730Z] ERROR: make.ps1 failed:
[2019-10-10T18:20:14.730Z] Unit tests failed
[2019-10-10T18:20:14.730Z] At C:\gopath\src\github.com\docker\docker\hack\make.ps1:324 char:32
[2019-10-10T18:20:14.730Z] +     if ($LASTEXITCODE -ne 0) { Throw "Unit tests failed" }
[2019-10-10T18:20:14.730Z] +                                ~~~~~~~~~~~~~~~~~~~~~~~~~
[2019-10-10T18:20:14.730Z] 
```

It could be a compile error, and if I remember correctly, Windows RS1 **does** print the actual failure (using the exact same script 🤷‍♂); I'm gonna trigger a rebuild with RS1 enabled to see if I can find what's wrong.

Probably also open a ticket for it, so that someone with more Windows knowledge could help fixing that

**Comment by thaJeztah** [comment]
There you go; Windows RS1 correctly shows the compile error, whereas RS5 just hides it (and both report all unit tests to have "passed", but possibly the problem occurs after that);

```
testutil\daemon\daemon.go:221:25: cannot use d (type *Daemon) as type string in argument to cleanupNetworkNamespace
```

**Comment by thaJeztah** [comment]
all green now

@cpuguy83 @tiborvass PTAL

**Review by tiborvass — APPROVED** [review]
LGTM but we should probably also add more `t.Helper()` in `testutil/daemon` (when daemons are starting/stopping the file lines are completely useless)

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
> but we should probably also add more t.Helper() in testutil/daemon (when daemons are starting/stopping the file lines are completely useless)

agreed; I didn't go looking for other ones yes; this one just stood out while testing my changes
